### PR TITLE
[DoctrineBridge] Fix required guess of boolean fields

### PR DESCRIPTION
**Pull request by enumag**
| Q | A |
| --- | --- |
| Bug fix? | yes |
| New feature? | no |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets |  |
| License | MIT |
| Doc PR |  |

### CODE REVIEW

**Comment by enumag** [comment]
Currently boolean fields of Doctrine entities are correctly guessed as checkboxes but incorrectly guessed as required.

File: src/Symfony/Bridge/Doctrine/Form/DoctrineOrmTypeGuesser.php [hunk 1]
```diff
@@ -95,7 +95,7 @@ public function guessRequired($class, $property)
 
         // Check whether the field exists and is nullable or not
         if ($classMetadata->hasField($property)) {
-            if (!$classMetadata->isNullable($property)) {
+            if (!$classMetadata->isNullable($property) && $classMetadata->getTypeOfField($property) !== 'boolean') {
```
Comment (Tobion) [inline src/Symfony/Bridge/Doctrine/Form/DoctrineOrmTypeGuesser.php]: please use `'boolean' !== $classMetadata->getTypeOfField($property)`

Comment (Tobion) [inline src/Symfony/Bridge/Doctrine/Form/DoctrineOrmTypeGuesser.php]: actually also use the constant please: https://github.com/doctrine/dbal/blob/master/lib/Doctrine/DBAL/Types/Type.php#L40

Comment (enumag) [inline src/Symfony/Bridge/Doctrine/Form/DoctrineOrmTypeGuesser.php]: Right... Forgot about the coding style.

**Comment by Tobion** [comment]
Please also refactor guessType to use DBAL constants.

File: src/Symfony/Bridge/Doctrine/Form/DoctrineOrmTypeGuesser.php [hunk 2]
```diff
@@ -51,28 +52,28 @@ public function guessType($class, $property)
         }
 
         switch ($metadata->getTypeOfField($property)) {
-            case 'array':
+            case Type::TARRAY:
```
Comment (enumag) [inline src/Symfony/Bridge/Doctrine/Form/DoctrineOrmTypeGuesser.php]: Should I add SIMPLE_ARRAY and JSON_ARRAY?

File: src/Symfony/Bridge/Doctrine/Form/DoctrineOrmTypeGuesser.php [hunk 3]
```diff
@@ -51,28 +52,28 @@ public function guessType($class, $property)
...
                 return new TypeGuess('collection', array(), Guess::MEDIUM_CONFIDENCE);
-            case 'boolean':
+            case Type::BOOLEAN:
                 return new TypeGuess('checkbox', array(), Guess::HIGH_CONFIDENCE);
-            case 'datetime':
+            case Type::DATETIME:
+            case Type::DATETIMETZ:
             case 'vardatetime':
```
Comment (enumag) [inline src/Symfony/Bridge/Doctrine/Form/DoctrineOrmTypeGuesser.php]: Not sure why this is here. There doesn't seem to be corresponding type in Doctrine.

File: src/Symfony/Bridge/Doctrine/Form/DoctrineOrmTypeGuesser.php (hunk 2 above)
Comment (Tobion) [inline src/Symfony/Bridge/Doctrine/Form/DoctrineOrmTypeGuesser.php]: Not sure if they work as collection form type as well. So lets do that in another PR if necessary.

File: src/Symfony/Bridge/Doctrine/Form/DoctrineOrmTypeGuesser.php (hunk 3 above)
Comment (Tobion) [inline src/Symfony/Bridge/Doctrine/Form/DoctrineOrmTypeGuesser.php]: https://github.com/doctrine/dbal/blob/master/lib/Doctrine/DBAL/Types/VarDateTimeType.php

**Comment by Tobion** [comment]
:+1: Should be merged in 2.3 by merger!

Status: Reviewed

**Comment by enumag** [comment]
@Tobion Thanks! I refactored guessPattern and guessMaxLength to use the constants as well.

**Comment by fabpot** [comment]
Thank you @enumag.
